pkg/ebpf/debugfs_stat_collector_linux.go

```diff
@@ -0,0 +1,86 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build linux_bpf
+
+package ebpf
+
+import (
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/DataDog/ebpf-manager/tracefs"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+const kProbeTelemetryName = "ebpf__probes"
+
+type DebugFsStatCollector struct {
+	hits           *prometheus.Desc
+	misses         *prometheus.Desc
+	lastProbeStats map[string]int
+}
+
+func NewDebugFsStatCollector() *DebugFsStatCollector {
+	return &DebugFsStatCollector{
+		hits:   prometheus.NewDesc(kProbeTelemetryName+"__hits", "Gauge tracking number of kprobe hits", []string{"name", "probe_type"}, nil),
+		misses: prometheus.NewDesc(kProbeTelemetryName+"__misses", "Gauge tracking number of kprobe misses", []string{"name", "probe_type"}, nil),
+	}
+}
+
+func (c *DebugFsStatCollector) updateProbeStats(pid int, probeType string, ch chan<- prometheus.Metric) {
+	if pid == 0 {
+		pid = myPid
+	}
+
+	root, err := tracefs.Root()
+	if err != nil {
+		log.Debugf("error getting tracefs root path: %s", err)
+		return
+	}
+	profile := filepath.Join(root, probeType+"_profile")
+	m, err := readKprobeProfile(profile)
+	if err != nil {
+		log.Debugf("error retrieving probe stats: %s", err)
+		return
+	}
+
+	for event, st := range m {
+		parts := eventRegexp.FindStringSubmatch(event)
+		if len(parts) > 2 {
+			// only get stats for our pid
+			if len(parts) > 3 {
+				parsePid, err := strconv.ParseInt(parts[3], 10, 32)
+				if err != nil || int(parsePid) != pid {
+					continue
+				}
+			}
+			// strip UID and PID from name
+			event = parts[1]
+		}
+		event = strings.ToLower(event)
+		hitsKey := "h_" + event
+		missesKey := "m_" + event
```

Comment:
The keys don't include the probe type, so you could have a collision between kprobe and uprobe.

Reply:
updated to include this